### ISSUE TITLE
Mirror the raycast/extensions repostiory and publish them automatically to the package registry

### ISSUE THREAD

**Issue by Yukaii** [BlastLauncher/blast#11]
## Steps

- TBD ~~Use a workflow to sync repository to the forked repository `blast/extension`~~

## Possible solutions...

The Raycast team runs the "publish" action with every new commit from branch `main`. Since it is a built-in action event on GitHub and run by the workflow inside the repository, they can ensure that the action runs with every commit. [^1]

If I manually or automatically mirror the raycast/extensions repository, the synchronization status (i.e., the last commit that was synced) must be stored somewhere, and generating the differences (i.e., determining which extensions need to be updated) will be more complex.

Alternatively, if I fork the raycast/extensions repository and replace the actions it runs, and handle git merging within the GitHub action context, I would still need to track the last synced commit [^2], then push each commit in between one by one to the forked repository, ensuring that every "publish" action is enqueued. ~~Additionally, if the synchronization process is run periodically as a workflow, I would need to ensure that only one synchronization workflow action is running at a time.~~ [^3]

[^1]: However, there may be a race condition where, for example, commit 1 changes extension A and commit 2 also changes extension A. In this case, the publish action triggered by commit 1 might run after the action triggered by commit 2, causing the latest published extension to be reverted back to commit 1. But since pull requests are manually merged, this scenario is unlikely to occur. (maybe?)

[^3]: [GitHub Actions supports only run one job ](https://github.blog/changelog/2021-04-19-github-actions-limit-workflow-run-or-job-concurrency/) at a time, so it will be fine

[^2]: Can it be stored as JSON [artifacts](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts)?

**Comment by Yukaii** [comment]
TLDR: 貫徹免費仔的架構需求好難設計啊

**Comment by Yukaii** [comment]
## Other Ideas

### Just zip the ray build result all at once to a repository and download them?

- Hard to implement upgrade/downgrade of an extension.
- How about metadata extraction?

**Comment by Yukaii** [comment]
### GitHub CLI playground

```bash
gh api \
  -H "Accept: application/vnd.github+json" \
  "/orgs/BlastLauncher/packages?package_type=npm"



# It takes 0.8 seconds to load, not sure if it's due to the network delay...
gh api \
  -H "Accept: application/vnd.github+json" \
    "/orgs/BlastLauncher/packages/npm/todo-list/versions"
```

So there no API to list all packages with versions Q_Q. For batch upgrading extensions, I'll still need one API server to cache cache packages states on GitHub.

**Comment by Yukaii** [comment]
## Requirements

Build a CLI that can meet [ray-cli action](https://github.com/raycast/github-actions/blob/2f74efbb1dde6c769599e067b0c18599525ec40c/ray/action.yml#L59-L68) needs, and a GitHub action that utilizes the CLI.

## Links

Some resources maybe useful when creating the CLI with nodejs

- [pkg](https://github.com/vercel/pkg) and [pkg-action](https://github.com/lando/pkg-action)
- [ncc](https://github.com/vercel/ncc)

### Note

In the first thread, I said **_I would still need to track the last synced commit_**, I found that it's not needed anymore, because in [ray action](https://github.com/raycast/github-actions/blob/2f74efbb1dde6c769599e067b0c18599525ec40c/get-paths/action.yml#L27-L33) they use [tj-actions/changed-files](https://github.com/tj-actions/changed-files) that has a `since_last_remote_commit` option. But we still need to deal with the merging the commits to the fork.

**Comment by Yukaii** [comment]
The first part is completed at [`sync-upstream.yml`](https://github.com/BlastLauncher/actions/blob/732135964c596ee07ef6e8b17d5b5f8818ba18ca/.github/workflows/sync-upstream.yml).

Second part is about to build a `blast-cli` that can build and changed raycast extension to GitHub Packages.
